pex/cli/commands/docs.py

```diff
@@ -0,0 +1,220 @@
+
+    @classmethod
+    def load(cls):
+        # type: () -> Optional[Pidfile]
+        try:
+            with open(cls._PIDFILE) as fp:
+                data = json.load(fp)
+            return cls(url=data["url"], pid=data["pid"])
+        except (OSError, IOError, ValueError, KeyError) as e:
+            logger.warning(
+                "Failed to load {server} pid file from {path}: {err}".format(
+                    server=SERVER_NAME, path=cls._PIDFILE, err=e
+                )
+            )
+            return None
+
+    @staticmethod
+    def _read_url(
+        server_log,  # type: str
+        timeout,  # type: float
+    ):
+        # type: (...) -> Optional[str]
+
+        # The permutations of Python versions, simple http server module and the output it provides:
+        #  2.7: Serving HTTP on 0.0.0.0 port 46399 ... -mSimpleHttpServer
+        #  3.5: Serving HTTP on 0.0.0.0 port 45577 ... -mhttp.server
+        # 3.6+: Serving HTTP on 0.0.0.0 port 33539 (http://0.0.0.0:33539/) ... -mhttp.server
+
+        start = time.time()
+        while time.time() - start < timeout:
+            with open(server_log) as fp:
+                for line in fp:
+                    if line.endswith(os.linesep):
+                        match = re.search(r"Serving HTTP on 0.0.0.0 port (?P<port>\d+)", line)
+                        if match:
+                            port = match.group("port")
+                            return "http://localhost:{port}".format(port=port)
+        return None
+
+    @classmethod
+    def record(
+        cls,
+        server_log,  # type: str
+        pid,  # type: int
+        timeout=5.0,  # type: float
+    ):
+        # type: (...) -> Optional[Pidfile]
+        url = cls._read_url(server_log, timeout)
+        if not url:
+            return None
+
+        with safe_open(cls._PIDFILE, "w") as fp:
+            json.dump(dict(url=url, pid=pid), fp, indent=2, sort_keys=True)
+        return cls(url=url, pid=pid)
+
+    url = attr.ib()  # type: str
+    pid = attr.ib()  # type: int
+
+    def alive(self):
+        # type: () -> bool
+        # TODO(John Sirois): Handle pid rollover
+        try:
+            os.kill(self.pid, 0)
+            return True
+        except OSError as e:
+            if e.errno == errno.ESRCH:  # No such process.
+                return False
+            raise
+
+    def kill(self):
+        # type: () -> None
+        os.kill(self.pid, signal.SIGTERM)
+
+
+@attr.s(frozen=True)
+class LaunchResult(object):
+    url = attr.ib()  # type: str
+    already_running = attr.ib()  # type: bool
+
+
+def launch_docs_server(
+    document_root,  # type: str
+    port,  # type: int
+    timeout=5.0,  # type: float
+):
+    # type: (...) -> Union[str, LaunchResult]
+
+    pidfile = Pidfile.load()
+    if pidfile and pidfile.alive():
+        return LaunchResult(url=pidfile.url, already_running=True)
+
+    # Not proper daemonization, but good enough.
```

Comment:
Did you consider the option of running in the foreground and letting the user daemonise themselves if they want (e.g. `pex3 docs &`)?

Reply:
Nope. This seemed like a good minimal feature set to start with. There are a few possible expansions. That is one. --browser is another for when open / xdg-open aren't available or configured right, etc.

Reply:
Ah, sorry I was a bit unclear, I was wondering about foreground-only rather than an option between foreground or background. That seems potentially being simpler to implement/manage (not having to do any subprocesses/output parsing, pidfiles etc), but you've got this working, so all good.

Reply:
It was already working really solid over a long burn in for `--devpi` which I've used for many months now. Next is to de-dup code there.

Reply:
I shored up the loose ends here in #2365 - I did not add a foreground mode yet though. I'm not sure if I want to commit to that until someone requests it for some reason.